src/document-click.js

```diff
@@ -358,3 +347,52 @@ export function expandTargetHref_(e, target, urlReplacements) {
   }
   return newHref;
 }
+
+
+/** @typedef {{element: !Element, attributes: Object<string, string>}} */
+let RestoreElementAttributesDef;
+
+
+/**
+ * Removes attributes from elements with matching hash target. This includes:
+ *   *[id=hash] and a[name=hash].
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+ * @param {string} hash
+ * @return {!Array<!RestoreElementAttributesDef>}}
+ * @private
+ */
+function removeAttrsWithMatchingHash_(ampdoc, hash) {
+  const restoreElementsAttrs = [];
+  const targetElements = ampdoc.getRootNode().querySelectorAll(
+      `#${hash},a[name=${hash}]`) || [];
+  for (let i = 0; i < targetElements.length; i++) {
+    const element = targetElements[i];
+    const attributes = {};
```

Comment:
I think doing loop here is a bit too much: we should just focus on a most likely case here and leave edge-cases out. The fidelity requirement is somewhat lower here.